src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -739,6 +739,67 @@ class NBDWatchCtx : public librbd::UpdateWatchCtx
   librados::IoCtx &io_ctx;
   librbd::Image &image;
   unsigned long size;
+  std::thread notify_thread;
+  ceph::condition_variable cond;
+  ceph::mutex lock = ceph::make_mutex("NBDWatchCtx::Locker");
+  bool notify = false;
+  bool terminated = false;
+
+  bool wait_notify() {
+    dout(10) << __func__ << dendl;
+
+    std::unique_lock locker{lock};
+    cond.wait(locker, [this] { return notify || terminated; });
+
+    notify = false;
+    if (terminated) {
+      return false;
+    }
+
+    dout(10) << __func__ << ": got notify request" << dendl;
+    return true;
+  }
+
+  void notify_entry() {
+    dout(10) << __func__ << dendl;
+
+    while (wait_notify()) {
+      librbd::image_info_t info;
```

Comment:
By switching from `image.stat()` to `image.size()`, we could ditch `info` and get size directly into `new_size`.